zkevm-circuits/src/bytecode_circuit/bytecode_unroller.rs

```diff
@@ -307,38 +307,38 @@ impl<F: Field> Config<F> {
         mut layouter: impl Layouter<F>,
         size: usize,
         witness: &[UnrolledBytecode<F>],
-    ) {
+    ) -> Result<(), Error> {
         let push_rindex_is_zero_chip = IsZeroChip::construct(self.push_rindex_is_zero.clone());
 
         // Subtract the unusable rows from the size
         let last_row_offset = size - self.minimum_rows + 1;
 
-        layouter
-            .assign_region(
-                || "assign bytecode",
-                |mut region| {
-                    let mut offset = 0;
-                    let mut push_rindex_prev = 0;
-
-                    for bytecode in witness.iter() {
-                        // Run over all the bytes
-                        let mut push_rindex = 0;
-                        let mut hash_rlc = F::zero();
-                        let hash_length = F::from(bytecode.bytes.len() as u64);
-                        for row in bytecode.rows.iter() {
-                            // Track which byte is an opcode and which is push
-                            // data
-                            let is_code = push_rindex == 0;
-                            let byte_push_size = get_push_size(row.byte.get_lower_128() as u8);
-                            push_rindex = if is_code {
-                                byte_push_size
-                            } else {
-                                push_rindex - 1
-                            };
-
-                            // Add the byte to the accumulator
-                            hash_rlc = hash_rlc * self.r + row.byte;
-
+        layouter.assign_region(
+            || "assign bytecode",
+            |mut region| {
+                let mut offset = 0;
+                let mut push_rindex_prev = 0;
+
+                for bytecode in witness.iter() {
+                    // Run over all the bytes
+                    let mut push_rindex = 0;
+                    let mut hash_rlc = F::zero();
+                    let hash_length = F::from(bytecode.bytes.len() as u64);
+                    for row in bytecode.rows.iter() {
+                        // Track which byte is an opcode and which is push
+                        // data
+                        let is_code = push_rindex == 0;
+                        let byte_push_size = get_push_size(row.byte.get_lower_128() as u8);
+                        push_rindex = if is_code {
+                            byte_push_size
+                        } else {
+                            push_rindex - 1
+                        };
+
+                        // Add the byte to the accumulator
+                        hash_rlc = hash_rlc * self.r + row.byte;
+
+                        if offset <= last_row_offset {
```

Comment:
If we don't enter this condition we know we're generating an invalid proof.  This is good for testing but it's not good outside of testing.
We have a similar situation in the state circuit, where we want to assign invalid advices for testing purposes, but we want an error in the assign function outside of tests.  In that case there's a generic const bool called [`SANITY_CHECK`](https://github.com/appliedzkp/zkevm-circuits/blob/a9bdc9a87f9fcb8068fd4a3f254c94194a0f4e1d/zkevm-circuits/src/state_circuit/state.rs#L554) that when true, makes the assign function panic on invalid witness data.  Could we have a similar implementation here?

Expanding on that, we could have two methods: `assign_unchecked` and `assign`. where the first one doesn't do sanity checks and the later does.

Reply:
My idea of `assign_unchecked` and `assign` was to do the following:
```rust
    fn _assign<const SANITY_CHECK: bool>(
        &self,
        mut layouter: impl Layouter<F>,
        size: usize,
        witness: &[UnrolledBytecode<F>],
    ) -> Result<(), Error> {
        // Full Implementation with conditional sanity check
    }
    pub(crate) fn assign(
        &self,
        mut layouter: impl Layouter<F>,
        size: usize,
        witness: &[UnrolledBytecode<F>],
    ) -> Result<(), Error> {
        self._assign::<true>(layouter, size, witness)
    }
    fn assign_unchecked(
        &self,
        mut layouter: impl Layouter<F>,
        size: usize,
        witness: &[UnrolledBytecode<F>],
    ) -> Result<(), Error> {
        self._assign::<false>(layouter, size, witness)
    }
```

But I like your idea of using the `#[cfg(test)]`!  Nevertheless, using features doesn't allow us to test both options in unit tests, so we wouldn't be able to test the sanity checks in our unit test functions.  But I think that's OK!

Reply:
I changed my mind a little. It really creates me some panic to have different code in production and test, and since we can activate features on tests, a better option could be:

```rust
let set_row = if cfg!(feature = "sanity_check") {
   offset <= last_row_offset
} else {
   true
};

if set_row {
   self.set_row(...
```

then in tests we could add 

```rust
/// Tests a circuit with incomplete bytecode (sanity check)
fn test_bytecode_incomplete() {
   let k = 9;
   let r = MyCircuit::r();
   verify::<Fr>(k, vec![unroll(vec![7u8; 2usize.pow(k) + 1], r)], false);
}

#[test]
#[cfg(feature = "sanity_check")]
fn bytecode_incomplete() {
   test_bytecode_incomplete();
}
#[test]
#[cfg(not(feature = "sanity_check"))]
#[should_panic]
fn bytecode_incomplete() {
   test_bytecode_incomplete();
}
```

then we can run the tests, activating the feature or not `cargo test --features=sanity_check`. The bad thing is that we have to run the tests twice.

Reply:
> then we can run the tests, activating the feature or not `cargo test --features=sanity_check`. The bad thing is that we have to run the tests twice.

Not just this; we also would be introducing and exposing a feature to the library that is only meaningful for internal tests.

Also from your example, from my point of view, when the sanity check is enabled, an invalid witness (or incompatible circuit setup) should cause an error or panic.  So to me it should look like:
```rust
if cfg!(feature = "sanity_check") {
  if offset > last_row_offset {
    panic!("offset > last_row_offset");
  }
}
if offset <= last_row_offset {
  self.set_row(...
}

Reply:
@ed255, with this code it will panic with "low k" is sanity_check is not set, and will panic with "offet > last_row_offset" if sanity_check is set.

I think that the main idea here, and (as I undestood from https://github.com/appliedzkp/zkevm-circuits/pull/403#discussion_r844088585) is how to create bad witnesses in a very targeted way (so, not using a generic fuzzer) to make the circuit to fail.

- Maybe we could decide if we introduce this kind of tests ( bad witnesses ) or not in our current state.
- Also (I think that we talked on this) is if we have to have a mechanism to decide when returning an error or when to panic ( to get the stacktrace ), because it also affects how "bad tests" are going yo be written. A would like some kind of retpanic!( Result::Err ) that panics or returns depending on compliation params.

Reply:
After an offline discussion with @adria0 we agree that it's very necessary to test the circuits with invalid inputs, but the fact that the `assign` functions are designed to calculate valid intermediate witness makes it hard, so we'd like to think more about this and move on with this PR for now (specially considering that the bytecode circuit has a [pending refactor](https://github.com/appliedzkp/zkevm-specs/issues/151)).

Reply:
I see that there's currently an open issue for negative tests on the circuits, maybe we can continue the discussion there? https://github.com/appliedzkp/zkevm-circuits/issues/400